### PR TITLE
[WIP] Try LTO again

### PR DESCRIPTION
**Pull request by nealrichardson**

### CODE REVIEW

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by nealrichardson** [comment]
Ok, that builds and links but segfaults on load. @jonkeane could you add back in here your cmake changes (I think you force-pushed them out of your previous branch) and try one more time?

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

This is not a complete solution (e.g. won't work well on non-LTO platforms), but should test if we *can* build on LTO

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

This isn't a complete solution (i.e. it likely will not work on non-lto builds), but should be enough to see if we can get through LTO.

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

Still not totally general, but another try.

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

File: r/tools/nixlibs.R [hunk 1]
```diff
@@ -328,20 +328,25 @@ build_libarrow <- function(src_dir, dst_dir) {
...
     RANLIB = R_CMD_config("RANLIB"),
-    LDFLAGS = R_CMD_config("LDFLAGS"),
-    CMAKE_CXX_COMPILER_AR = R_CMD_config("LDFLAGS")
+    LDFLAGS = R_CMD_config("LDFLAGS")
   )
   env_vars <- paste0(names(env_var_list), '="', env_var_list, '"', collapse = " ")
   env_vars <- with_s3_support(env_vars)
   env_vars <- with_mimalloc(env_vars)
-  if (tolower(Sys.info()[["sysname"]]) %in% "sunos") {
+  if (tolower(Sys.info()[["sysname"]]) %in% "sunos" || grepl("-flto", env_var_list[["LDFLAGS"]])) {
```
Comment (jonkeane) [inline r/tools/nixlibs.R:336]: If you wanted to be slightly more targeted you could also check for `"-flto"` in `R CMD config LTO`

Comment (nealrichardson) [reply r/tools/nixlibs.R:336]: Maybe, but I think this is where it matters, this is what gets used in the link step right?

Comment (nealrichardson) [reply r/tools/nixlibs.R:336]: Hmm, `-flto` doesn't show up in LDFLAGS: `LDFLAGS="-L/usr/local/lib"`

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

File: r/tools/nixlibs.R [hunk 2]
```diff
@@ -326,20 +326,27 @@ build_libarrow <- function(src_dir, dst_dir) {
     CC = R_CMD_config("CC"),
     CXX = paste(R_CMD_config("CXX11"), R_CMD_config("CXX11STD")),
     # CXXFLAGS = R_CMD_config("CXX11FLAGS"), # We don't want the same debug symbols
-    ARROW_R_CXXFLAGS = paste(Sys.getenv("ARROW_R_CXXFLAGS", ""), "-fno-lto"),
-    LDFLAGS = R_CMD_config("LDFLAGS")
+    AR = R_CMD_config("AR"),
+    RANLIB = R_CMD_config("RANLIB"),
+    LDFLAGS = paste(R_CMD_config("LDFLAGS"), R_CMD_config("LTO_LD"))
```
Comment (nealrichardson) [inline r/tools/nixlibs.R:331]:
```suggestion
    LDFLAGS = paste(R_CMD_config("LDFLAGS"), R_CMD_config("LTO"))
```

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -41,6 +41,11 @@ if(POLICY CMP0068)
   cmake_policy(SET CMP0068 NEW)
 endif()
 
+if(POLICY CMP0069)
+  # https://cmake.org/cmake/help/latest/policy/CMP0069.html
+  cmake_policy(SET CMP0069 OLD)
+endif()
+
```
Comment (nealrichardson) [inline cpp/CMakeLists.txt:48]:
```suggestion
```

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-rhub-debian-gcc-devel-lto-latest

**Pull request closed without merging**